CE-326

denizgenc opened the pull request:
See ticket for a (too) detailed explanation of the work done for this, though obviously I welcome any questions, comments and change requests.

Creating a draft PR until I can watch all the lambdas trigger normally at 4 p.m. BST (in about 45 minutes time), to ensure there are no issues (the lambda now passes all tests, so there should not be any).

denizgenc committed:
Break out of `get_group_info()` if rate limit exceeded

This took me way too long because Google's documentation for this
library is crap. These are the official docs for the `HttpError` class:
https://googleapis.github.io/google-api-python-client/docs/epy/googleapiclient.errors.HttpError-class.html

Absolutely no details on the different fields provided by the class
(which I need to search for "exceeded" in the error response). Had to
look into the actual implementation:
https://github.com/googleapis/google-api-python-client/blob/d5cf4e0b1537eeafdb96ae831bbdffc5f1c6a94c/googleapiclient/errors.py#L35

(also, HttpError.reason is generated using a different method to the the
"reason" in _should_retry_response() function, which is run as part of
API calls and logs warnings:
https://github.com/googleapis/google-api-python-client/blob/cbb1f88b82b21f5cb9dcace33ffea3f95a189015/googleapiclient/http.py#L139
This leads to the logging output saying "rateLimitExceeded" while
HttpError's string representation contains the more detailed "Quota
exceeded for quota metric ..." - really confusing)

denizgenc committed:
Correct the name of the Cloudwatch event rule and trigger

denizgenc committed:
Fix rate limit test

v1.10 of the API does not provide the `reason` field as part of
HttpError - it's only created when HttpError's `__repr__()` is called.

denizgenc committed:
Fix format string mistake

argh

denizgenc committed:
Move the lambda execution role into this repository

This is in preparation for adding the "lambda:InvokeFunction" permission
to the role so the Lambda itself can invoke more copies of itself to
process data concurrently.

The file terraform/modules/lambda-execution-role/lambda_iam.tf was
previously in our internal Terraform repository, where it was part of
the Concourse role previously used to deploy this service.

denizgenc committed:
Add InvokeFunction perms to lambda execution role

denizgenc committed:
Invoke Lambda for each page of Google Groups

This bypasses the issue where we would time out in the get_group_info()
for loop if there were too many groups to process serially.

Now we just invoke another Lambda for each page of Google Groups we get
from build_group_dict(), essentially parallelising it.

I decided to pass nextPageToken and lambda_function_name from function
to function until it got to build_group_dict(). I realise this can be a
bit confusing since these parameters aren't used at all in
print_group_info() beyond passing it on to build_group_dict(). The other
option was to make these values globals - I feel like this actually is a
good opportunity to use them, but I was worried that others may not
agree.

denizgenc committed:
Prettify code

denizgenc committed:
Update tests to reflect changes in script.py

We're essentially testing three separate invocations of the lambda now.
I moved the class out of the test function into the body to make it
easier to reference (and also changed its name to prevent aliasing (?)
the class inside test_get_group_info(), though that might not have been
necessary.

denizgenc committed:
Decorate groups method with @staticmethod

denizgenc commented on lambda/script.py:
```diff
@@ -140,15 +149,30 @@ def print_group_info(
 
 
-def main(event: Dict, context: Dict):
+def main(event: Dict, context: Any):
```
`context` is a "context object", not a dictionary. See https://docs.aws.amazon.com/lambda/latest/dg/python-context.html

Since `function_name` is accessed like an object property, instead of like a dictionary key (because... it *is* an object property), line 167 would've raised a type hinting error since `Dict` instances do not have a `function_name` property.

From a cursory read of the documentation for the `typing` library, it seems that `Any` is the best type hint for "arbitrary" objects where we make use of their fields or functions, see https://docs.python.org/3.7/library/typing.html#the-any-type

denizgenc commented:
Had a look at the most recent lambda runs. Everything's working without problems, so I'm marking this as ready for review

alice-carr commented on lambda/script.py:
```diff
@@ -121,6 +125,9 @@ def get_group_info(
         except HttpError as e:
             print(f"Http error for group {group_id}: {e}")
+            if "exceeded" in f"{e}".lower():
+                print("Rate limit exceeded, stopping loop over groups...")
```
are we likely to exceed our rate limit? If so, we're just exiting out -  does this mean we're losing data?

denizgenc replied:
We could be, and in fact, we probably are... I just had a look at the CloudWatch logs and since this change the lambda for the Cabinet Office domain has been emitting a new error:
`"Quota exceeded for quota metric 'Queries' and limit 'Queries per minute per user' of service 'groupssettings.googleapis.com' for consumer ...`

Previously, we'd get an error for hitting the quota for 'Queries per day', which meant that for several hours of each day every lambda invocation would fail until that quota would reset. Now (because we’re executing multiple lambdas at the same time), we seem to hit a *new* limit (the per minute request quota), and at that point stop getting Google Group data and start logging what we've collected up to then.

I believe that if we fix this 'Queries per minute' issue, we'll be hitting the 'Queries per day' error again. Honestly, that would be preferable, because right now we're getting *incomplete* information on a regular basis, while we used to get *complete* (I think) information, just less frequently than we wanted (about once or twice a day instead of every hour).

Until the quota limits are raised, I don't think we'll be getting the data we want (“the data we want” being: a description of how *all* the Google Groups in a domain are configured at the start of every hour). I think we need to talk to several people:
- Those that are expecting the Google Groups data being in Splunk on time and in full, to warn them of potential gaps in the data. We should also ask them whether having this data about the Google Groups every hour is really what they need or if it would be OK to have it on a less frequent schedule.
- Those that can request a quota limit increase from Google on the CO domain. We can probably figure out what both the 'Queries per minute' limit should be, given an API call takes about 180ms, and the 'Queries per day' limit, as we know how many Groups there are in the domain and that we're making a request for each one 24 times a day.

alice-carr reviewed (dismissed):
Hey Deniz this all looks good - I've asked one question above though

denizgenc committed:
WIP code change

- Move lambda invocation to the end
  - This means we can remove the function name but we had to refactor a
    little to bubble up the lambda_payload
- Use the num_retries argument on
  google_client.groups().get().execute() to automatically retry
  - This may or may not work, depending on whether groups() returns a
    HttpRequest object or a BatchHttpRequest object. It should work for
    HttpRequest, shouldn't for BatchHttpRequest.

    Based on this
    https://googleapis.github.io/google-api-python-client/docs/dyn/groupssettings_v1.html
    I think it returns a BatchHttpRequest object but I'm trying to
    remain positive
  - If this doesn't work, use the solution at
    https://stackoverflow.com/a/7663441/12833309 instead

denizgenc committed:
Update tests
